File: go/styles.go
package main

import "math"

// defaultFCLStyle returns the default FCL button style.
func defaultFCLStyle(name ...string) *OrderedMap {
	styleName := "Default"
	if len(name) > 0 {
		styleName = name[0]
	}
	return NewOrderedMapFromPairs(
		"name", styleName,
		"textColor", -1,
		"textSize", 12,
		"strokeColor", -12303292,
		"strokeWidth", 10,
		"cornerRadius", 100,
		"fillColor", 0,
		"textColorPressed", -1,
		"textSizePressed", 12,
		"strokeColorPressed", -12303292,
		"strokeWidthPressed", 10,
		"cornerRadiusPressed", 100,
		"fillColorPressed", -3355444,
	)
}

// defaultZLFallbackFCLStyle returns the ZL native default FCL style.
func defaultZLFallbackFCLStyle(name ...string) *OrderedMap {
	styleName := "ZL Native Default"
	if len(name) > 0 {
		styleName = name[0]
	}
	return NewOrderedMapFromPairs(
		"name", styleName,
		"textColor", -1,
		"textSize", 14,
		"strokeColor", -1,
		"strokeWidth", 0,
		"cornerRadius", 0,
		"fillColor", -2147483648,
		"textColorPressed", -1,
		"textSizePressed", 14,
		"strokeColorPressed", -1,
		"strokeWidthPressed", 0,
		"cornerRadiusPressed", 0,
		"fillColorPressed", -1282897784,
	)
}

// defaultFCLDirectionStyle returns the default FCL direction style.
func defaultFCLDirectionStyle() *OrderedMap {
	return NewOrderedMapFromPairs(
		"name", "Default",
		"styleType", "BUTTON",
		"buttonStyle", NewOrderedMapFromPairs(
			"interval", 50,
			"textColor", -1,
			"textSize", 12,
			"strokeColor", -12303292,
			"strokeWidth", 10,
			"cornerRadius", 100,
			"fillColor", 0,
			"textColorPressed", -1,
			"textSizePressed", 12,
			"strokeColorPressed", -12303292,
			"strokeWidthPressed", 10,
			"cornerRadiusPressed", 100,
			"fillColorPressed", -3355444,
		),
		"rockerStyle", NewOrderedMapFromPairs(
			"rockerSize", 400,
			"bgCornerRadius", 500,
			"bgStrokeWidth", 20,
			"bgStrokeColor", -12303292,
			"bgFillColor", 0,
			"rockerCornerRadius", 500,
			"rockerStrokeWidth", 10,
			"rockerStrokeColor", -12303292,
			"rockerFillColor", -7829368,
		),
	)
}

// emptyFCLEvent returns an empty FCL event.
func emptyFCLEvent() *OrderedMap {
	return NewOrderedMapFromPairs(
		"autoKeep", false,
		"autoClick", false,
		"openMenu", false,
		"switchTouchMode", false,
		"switchMouseMode", false,
		"input", false,
		"quickInput", false,
		"outputText", "",
		"outputKeycodes", []interface{}{},
		"bindViewGroup", []interface{}{},
	)
}

// fclButtonEvent returns the default FCL button event structure.
func fclButtonEvent() *OrderedMap {
	return NewOrderedMapFromPairs(
		"pointerFollow", false,
		"Movable", false,
		"pressEvent", emptyFCLEvent(),
		"longPressEvent", emptyFCLEvent(),
		"clickEvent", emptyFCLEvent(),
		"doubleClickEvent", emptyFCLEvent(),
	)
}

// zlShapeToFCLRadius converts ZL border radius shape to FCL corner radius.
func zlShapeToFCLRadius(shape interface{}) int {
	s, ok := asOrderedMap(shape)
	if !ok {
		return 100
	}
	keys := []string{"topStart", "topEnd", "bottomEnd", "bottomStart"}
	var values []float64
	for _, k := range keys {
		values = append(values, clampZLShape(getOr(s, k, 0.0), 0.0))
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return clampInt(sum/float64(len(values))*10, 100)
}

// defaultZLJoystickStyleConfig returns the default ZL joystick style config.
func defaultZLJoystickStyleConfig() *OrderedMap {
	return NewOrderedMapFromPairs(
		"alpha", PyFloat(1.0),
		"backgroundColor", fclARGBToZLColor(0x80000000),
		"joystickColor", fclARGBToZLColor(0x80FFFFFF),
		"joystickCanLockColor", fclARGBToZLColor(0x80FFFF00),
		"joystickLockedColor", fclARGBToZLColor(0x8000FF00),
		"lockMarkColor", fclARGBToZLColor(0xFFFFFFFF),
		"borderWidthRatio", 0,
		"borderColor", fclARGBToZLColor(0xFFFFFFFF),
		"backgroundShape", 50,
		"joystickShape", 50,
		"joystickSize", PyFloat(0.5),
	)
}

// fclRockerStyleToZLJoystick converts FCL rocker style to ZL joystick style.
func fclRockerStyleToZLJoystick(style *OrderedMap) *OrderedMap {
	var rocker *OrderedMap
	if style != nil {
		if r, ok := style.Get("rockerStyle"); ok {
			if rm, ok := r.(*OrderedMap); ok && rm.Len() > 0 {
				rocker = rm
			}
		}
	}
	if rocker == nil {
		rocker = NewOrderedMap()
	}
	config := defaultZLJoystickStyleConfig()
	config.Set("backgroundColor", fclARGBToZLColor(getOr(rocker, "bgFillColor", 0x80000000)))
	config.Set("joystickColor", fclARGBToZLColor(getOr(rocker, "rockerFillColor", 0x80FFFFFF)))
	config.Set("borderColor", fclARGBToZLColor(getOr(rocker, "bgStrokeColor", 0xFFFFFFFF)))
	config.Set("borderWidthRatio", maxInt(0, minInt(50, clampInt(getOr(rocker, "bgStrokeWidth", 0))/10)))
	config.Set("backgroundShape", fclRadiusToZLPercent(getOr(rocker, "bgCornerRadius", 500)))
	config.Set("joystickShape", fclRadiusToZLPercent(getOr(rocker, "rockerCornerRadius", 500)))
	config.Set("joystickSize", PyFloat(fclRatioToZL(getOr(rocker, "rockerSize", 500))))
	return NewOrderedMapFromPairs(
		"uuid", shortID(),
		"lightStyle", config,
		"darkStyle", deepCopyJSON(config),
	)
}

// directionStyleMap builds a name->style map from a list of direction styles.
func directionStyleMap(styles []interface{}) map[string]*OrderedMap {
	result := map[string]*OrderedMap{}
	for _, item := range styles {
		s, ok := item.(*OrderedMap)
		if !ok {
			continue
		}
		name := toString(getOr(s, "name", ""))
		result[name] = s
	}
	return result
}

// resolveDirectionStyle resolves the direction style from a direction object.
func resolveDirectionStyle(direction *OrderedMap, styles map[string]*OrderedMap) *OrderedMap {
	style, ok := direction.Get("style")
	if ok {
		if sm, ok := style.(*OrderedMap); ok {
			return sm
		}
	}
	if s, ok := styles[toString(style)]; ok {
		return s
	}
	return NewOrderedMap()
}

// styleNameForZLStyle generates a style name with a suffix.
func styleNameForZLStyle(baseName, uuidValue string) string {
	suffix := ""
	if uuidValue != "" {
		if len(uuidValue) >= 6 {
			suffix = uuidValue[:6]
		} else {
			suffix = uuidValue
		}
	} else {
		s := shortID()
		if len(s) >= 6 {
			suffix = s[:6]
		} else {
			suffix = s
		}
	}
	return "ZL " + baseName + " " + suffix
}

// fclStylesToZL converts FCL button styles to ZL styles.
// Returns (styles list, name->uuid mapping).
func fclStylesToZL(styles []interface{}) ([]interface{}, map[string]string) {
	if len(styles) == 0 {
		styles = []interface{}{defaultFCLStyle()}
	}
	result := []interface{}{}
	mapping := map[string]string{}
	for _, item := range styles {
		style, ok := item.(*OrderedMap)
		if !ok {
			continue
		}
		name := toString(getOr(style, "name", "Default"))
		if name == "" {
			name = "Default"
		}
		sid := shortID()
		mapping[name] = sid
		radius := clampZLShape(clampFloat(getOr(style, "cornerRadius", 0), 0.0) / 10.0)
		pressedRadius := clampZLShape(clampFloat(getOr(style, "cornerRadiusPressed", getOr(style, "cornerRadius", 0)), 0.0) / 10.0)
		light := NewOrderedMapFromPairs(
		"alpha", PyFloat(1.0),
		"pressedAlpha", PyFloat(1.0),
		"backgroundColor", fclARGBToZLColor(getOr(style, "fillColor", 0)),
		"pressedBackgroundColor", fclARGBToZLColor(getOr(style, "fillColorPressed", -3355444)),
		"contentColor", fclARGBToZLColor(getOr(style, "textColor", -1), -1),
		"pressedContentColor", fclARGBToZLColor(getOr(style, "textColorPressed", -1), -1),
		"fontSize", fclFontToZL(getOr(style, "textSize", 12)),
		"pressedFontSize", fclFontToZL(getOr(style, "textSizePressed", getOr(style, "textSize", 12))),
		"borderWidth", clampZLBorderWidth(clampInt(getOr(style, "strokeWidth", 10))/10),
		"pressedBorderWidth", clampZLBorderWidth(clampInt(getOr(style, "strokeWidthPressed", 10))/10),
		"borderColor", fclARGBToZLColor(getOr(style, "strokeColor", -12303292), -12303292),
		"pressedBorderColor", fclARGBToZLColor(getOr(style, "strokeColorPressed", -12303292), -12303292),
		"borderRadius", NewOrderedMapFromPairs("topStart", PyFloat(radius), "topEnd", PyFloat(radius), "bottomEnd", PyFloat(radius), "bottomStart", PyFloat(radius)),
		"pressedBorderRadius", NewOrderedMapFromPairs("topStart", PyFloat(pressedRadius), "topEnd", PyFloat(pressedRadius), "bottomEnd", PyFloat(pressedRadius), "bottomStart", PyFloat(pressedRadius)),
	)
		result = append(result, NewOrderedMapFromPairs(
			"name", name,
			"uuid", sid,
			"animateSwap", false,
			"commonStyle", true,
			"lightStyle", light,
			"darkStyle", deepCopyJSON(light),
		))
	}
	return result, mapping
}

// makeZLButtonSize builds a ZL buttonSize object from FCL baseInfo.
func makeZLButtonSize(baseInfo *OrderedMap, absoluteAsPercentage bool, aspect float64) *OrderedMap {
	pw := getOrOrderedMap(baseInfo, "percentageWidth")
	ph := getOrOrderedMap(baseInfo, "percentageHeight")
	if toString(getOr(baseInfo, "sizeType", "")) == "ABSOLUTE" && absoluteAsPercentage {
		screenHeightDP := 411.0
		screenWidthDP := screenHeightDP * math.Max(0.1, clampFloat(aspect, 16.0/9.0))
		widthPercentage := maxInt(100, minInt(10000, int(math.Round(clampZLDP(getOr(baseInfo, "absoluteWidth", 50))/screenWidthDP*10000))))
		heightPercentage := maxInt(100, minInt(10000, int(math.Round(clampZLDP(getOr(baseInfo, "absoluteHeight", 50))/screenHeightDP*10000))))
		return NewOrderedMapFromPairs(
			"type", "percentage",
			"widthDp", PyFloat(clampZLDP(getOr(baseInfo, "absoluteWidth", 50))),
			"heightDp", PyFloat(clampZLDP(getOr(baseInfo, "absoluteHeight", 50))),
			"widthPercentage", widthPercentage,
			"heightPercentage", heightPercentage,
			"widthReference", "screen_width",
			"heightReference", "screen_height",
		)
	}
	sizeType := "percentage"
	if toString(getOr(baseInfo, "sizeType", "")) == "ABSOLUTE" {
		sizeType = "dp"
	}
	return NewOrderedMapFromPairs(
		"type", sizeType,
		"widthDp", PyFloat(clampZLDP(getOr(baseInfo, "absoluteWidth", 50))),
		"heightDp", PyFloat(clampZLDP(getOr(baseInfo, "absoluteHeight", 50))),
		"widthPercentage", fclSizeToZL(getOr(pw, "size", 50)),
		"heightPercentage", fclSizeToZL(getOr(ph, "size", 50)),
		"widthReference", fclRefToZL(toString(getOr(pw, "reference", ""))),
		"heightReference", fclRefToZl(toString(getOr(ph, "reference", ""))),
	)
}

// fclRefToZl converts FCL reference name to ZL reference name.
func fclRefToZl(reference string) string {
	if reference == "SCREEN_HEIGHT" {
		return "screen_height"
	}
	return "screen_width"
}

File: cc-rs/src/styles.rs
use crate::utils::*;
use serde_json::{json, Map, Value};
use std::collections::HashMap;

pub fn default_fcl_style(name: Option<&str>) -> Value {
    let style_name = name.unwrap_or("Default");
    json!({
        "name": style_name,
        "textColor": -1,
        "textSize": 12,
        "strokeColor": -12303292,
        "strokeWidth": 10,
        "cornerRadius": 100,
        "fillColor": 0,
        "textColorPressed": -1,
        "textSizePressed": 12,
        "strokeColorPressed": -12303292,
        "strokeWidthPressed": 10,
        "cornerRadiusPressed": 100,
        "fillColorPressed": -3355444,
    })
}

pub fn default_zl_fallback_fcl_style(name: Option<&str>) -> Value {
    let style_name = name.unwrap_or("ZL Native Default");
    json!({
        "name": style_name,
        "textColor": -1,
        "textSize": 14,
        "strokeColor": -1,
        "strokeWidth": 0,
        "cornerRadius": 0,
        "fillColor": -2147483648,
        "textColorPressed": -1,
        "textSizePressed": 14,
        "strokeColorPressed": -1,
        "strokeWidthPressed": 0,
        "cornerRadiusPressed": 0,
        "fillColorPressed": -1282897784,
    })
}

pub fn default_fcl_direction_style() -> Value {
    json!({
        "name": "Default",
        "styleType": "BUTTON",
        "buttonStyle": {
            "interval": 50,
            "textColor": -1,
            "textSize": 12,
            "strokeColor": -12303292,
            "strokeWidth": 10,
            "cornerRadius": 100,
            "fillColor": 0,
            "textColorPressed": -1,
            "textSizePressed": 12,
            "strokeColorPressed": -12303292,
            "strokeWidthPressed": 10,
            "cornerRadiusPressed": 100,
            "fillColorPressed": -3355444,
        },
        "rockerStyle": {
            "rockerSize": 400,
            "bgCornerRadius": 500,
            "bgStrokeWidth": 20,
            "bgStrokeColor": -12303292,
            "bgFillColor": 0,
            "rockerCornerRadius": 500,
            "rockerStrokeWidth": 10,
            "rockerStrokeColor": -12303292,
            "rockerFillColor": -7829368,
        },
    })
}

pub fn empty_fcl_event() -> Value {
    json!({
        "autoKeep": false,
        "autoClick": false,
        "openMenu": false,
        "switchTouchMode": false,
        "switchMouseMode": false,
        "input": false,
        "quickInput": false,
        "outputText": "",
        "outputKeycodes": [],
        "bindViewGroup": [],
    })
}

pub fn fcl_button_event() -> Value {
    json!({
        "pointerFollow": false,
        "Movable": false,
        "pressEvent": empty_fcl_event(),
        "longPressEvent": empty_fcl_event(),
        "clickEvent": empty_fcl_event(),
        "doubleClickEvent": empty_fcl_event(),
    })
}

pub fn zl_shape_to_fcl_radius(shape: &Value) -> i64 {
    let Some(s) = as_object(shape) else {
        return 100;
    };
    let keys = ["topStart", "topEnd", "bottomEnd", "bottomStart"];
    let mut sum = 0.0;
    for k in keys {
        let v = s.get(k).cloned().unwrap_or(Value::from(0));
        sum += clamp_zl_shape(&v, 0.0);
    }
    0.max(500.min(py_round(sum / keys.len() as f64 * 10.0)))
}

pub fn default_zl_joystick_style_config() -> Value {
    json!({
        "alpha": py_num(1.0),
        "backgroundColor": fcl_argb_to_zl_color_i(0x80000000),
        "joystickColor": fcl_argb_to_zl_color_i(0x80FFFFFF),
        "joystickCanLockColor": fcl_argb_to_zl_color_i(0x80FFFF00),
        "joystickLockedColor": fcl_argb_to_zl_color_i(0x8000FF00),
        "lockMarkColor": fcl_argb_to_zl_color_i(0xFFFFFFFF),
        "borderWidthRatio": 0,
        "borderColor": fcl_argb_to_zl_color_i(0xFFFFFFFF),
        "backgroundShape": 50,
        "joystickShape": 50,
        "joystickSize": py_num(0.5),
    })
}

fn set_config_from(
    config: &mut Value,
    src: &Value,
    fill_key: &str,
    fill_def: i64,
    text_key: &str,
    text_def: i64,
    stroke_key: &str,
    stroke_def: i64,
    stroke_width_key: &str,
    stroke_width_def: i64,
    corner_key: &str,
    corner_def: i64,
    joystick_corner_key: &str,
    joystick_corner_def: i64,
    joystick_size: f64,
) {
    if let Value::Object(c) = config {
        c.insert(
            "backgroundColor".to_string(),
            inum(fcl_argb_to_zl_color(&get_or(src, fill_key, inum(fill_def)), 0)),
        );
        c.insert(
            "joystickColor".to_string(),
            inum(fcl_argb_to_zl_color(&get_or(src, text_key, inum(text_def)), 0)),
        );
        c.insert(
            "borderColor".to_string(),
            inum(fcl_argb_to_zl_color(&get_or(src, stroke_key, inum(stroke_def)), 0)),
        );
        let w = clamp_int(&get_or(src, stroke_width_key, inum(stroke_width_def)), 0);
        c.insert("borderWidthRatio".to_string(), inum(0.max(50.min(w / 10))));
        c.insert(
            "backgroundShape".to_string(),
            inum(fcl_radius_to_zl_percent(&get_or(src, corner_key, inum(corner_def)), 500)),
        );
        c.insert(
            "joystickShape".to_string(),
            inum(fcl_radius_to_zl_percent(&get_or(src, joystick_corner_key, inum(joystick_corner_def)), 500)),
        );
        c.insert("joystickSize".to_string(), py_num(joystick_size));
    }
}

pub fn fcl_rocker_style_to_zl_joystick(style: Option<&Value>) -> Value {
    let mut rocker = obj();
    let mut name = "Default".to_string();
    if let Some(s) = style {
        name = to_string_v(&get_or(s, "name", json!("Default")));
        if let Some(r) = s.get("rockerStyle") {
            if let Value::Object(rm) = r {
                if !rm.is_empty() {
                    rocker = r.clone();
                }
            }
        }
    }
    let mut config = default_zl_joystick_style_config();
    set_config_from(
        &mut config,
        &rocker,
        "bgFillColor",
        0x80000000,
        "rockerFillColor",
        0x80FFFFFF,
        "bgStrokeColor",
        0xFFFFFFFF,
        "bgStrokeWidth",
        0,
        "bgCornerRadius",
        500,
        "rockerCornerRadius",
        500,
        fcl_ratio_to_zl(&get_or(&rocker, "rockerSize", inum(500)), 500),
    );
    json!({
        "name": name,
        "uuid": short_id(),
        "commonStyle": true,
        "lightStyle": config,
        "darkStyle": config.clone(),
    })
}

pub fn fcl_button_style_to_zl_joystick(style: Option<&Value>) -> Value {
    let mut btn = obj();
    let mut name = "Default".to_string();
    if let Some(s) = style {
        name = to_string_v(&get_or(s, "name", json!("Default")));
        if let Some(b) = s.get("buttonStyle") {
            if let Value::Object(bm) = b {
                if !bm.is_empty() {
                    btn = b.clone();
                }
            }
        }
    }
    let mut config = default_zl_joystick_style_config();
    set_config_from(
        &mut config,
        &btn,
        "fillColor",
        0x80000000,
        "textColor",
        0x80FFFFFF,
        "strokeColor",
        0xFFFFFFFF,
        "strokeWidth",
        10,
        "cornerRadius",
        100,
        "cornerRadius",
        100,
        0.5,
    );
    json!({
        "name": name,
        "uuid": short_id(),
        "commonStyle": true,
        "lightStyle": config,
        "darkStyle": config.clone(),
    })
}

pub fn direction_style_map<'a>(styles: &'a [Value]) -> HashMap<String, &'a Value> {
    let mut result = HashMap::new();
    for item in styles {
        if !item.is_object() {
            continue;
        }
        let name = to_string_v(&get_or(item, "name", json!("")));
        result.insert(name, item);
    }
    result
}

pub fn resolve_direction_style(
    direction: &Value,
    styles: &HashMap<String, &Value>,
) -> Value {
    match direction.get("style") {
        Some(style) => {
            if style.is_object() {
                return style.clone();
            }
            let name = to_string_v(style);
            styles.get(&name).map(|s| (*s).clone()).unwrap_or_else(obj)
        }
        None => styles
            .get("")
            .map(|s| (*s).clone())
            .unwrap_or_else(obj),
    }
}

pub fn style_name_for_zl_style(base_name: &str, uuid_value: &str) -> String {
    let suffix = if !uuid_value.is_empty() {
        if uuid_value.len() >= 6 {
            uuid_value[..6].to_string()
        } else {
            uuid_value.to_string()
        }
    } else {
        let s = short_id();
        if s.len() >= 6 {
            s[..6].to_string()
        } else {
            s
        }
    };
    format!("ZL {} {}", base_name, suffix)
}

pub fn fcl_styles_to_zl(styles: &[Value]) -> (Vec<Value>, HashMap<String, String>) {
    let default_style;
    let styles: &[Value] = if styles.is_empty() {
        default_style = vec![default_fcl_style(None)];
        &default_style
    } else {
        styles
    };
    let mut result = Vec::new();
    let mut mapping = HashMap::new();
    for item in styles {
        if !item.is_object() {
            continue;
        }
        let style = item;
        let mut name = to_string_v(&get_or(style, "name", json!("Default")));
        if name.is_empty() {
            name = "Default".to_string();
        }
        let sid = short_id();
        mapping.insert(name.clone(), sid.clone());
        let radius =
            clamp_zl_shape(&gof_num(clamp_float(&get_or(style, "cornerRadius", inum(0)), 0.0) / 10.0), 0.0);
        let pressed_src = get_or(style, "cornerRadiusPressed", get_or(style, "cornerRadius", inum(0)));
        let pressed_radius = clamp_zl_shape(&gof_num(clamp_float(&pressed_src, 0.0) / 10.0), 0.0);
        let stroke_width = clamp_int(&get_or(style, "strokeWidth", inum(10)), 0);
        let stroke_width_pressed = clamp_int(&get_or(style, "strokeWidthPressed", inum(10)), 0);
        let text_size = get_or(style, "textSize", inum(12));
        let radius_obj = {
            let mut m = Map::new();
            for k in ["topStart", "topEnd", "bottomEnd", "bottomStart"] {
                m.insert(k.to_string(), py_num(radius as f64));
            }
            Value::Object(m)
        };
        let pressed_radius_obj = {
            let mut m = Map::new();
            for k in ["topStart", "topEnd", "bottomEnd", "bottomStart"] {
                m.insert(k.to_string(), py_num(pressed_radius as f64));
            }
            Value::Object(m)
        };
        let light = json!({
            "alpha": py_num(1.0),
            "pressedAlpha": py_num(1.0),
            "backgroundColor": inum(fcl_argb_to_zl_color(&get_or(style, "fillColor", inum(0)), 0)),
            "pressedBackgroundColor": inum(fcl_argb_to_zl_color(&get_or(style, "fillColorPressed", inum(-3355444)), 0)),
            "contentColor": inum(fcl_argb_to_zl_color(&get_or(style, "textColor", inum(-1)), -1)),
            "pressedContentColor": inum(fcl_argb_to_zl_color(&get_or(style, "textColorPressed", inum(-1)), -1)),
            "fontSize": inum(fcl_font_to_zl(&text_size, 12)),
            "pressedFontSize": inum(fcl_font_to_zl(&get_or(style, "textSizePressed", text_size.clone()), 12)),
            "borderWidth": inum(clamp_zl_border_width(&inum(stroke_width / 10), 0)),
            "pressedBorderWidth": inum(clamp_zl_border_width(&inum(stroke_width_pressed / 10), 0)),
            "borderColor": inum(fcl_argb_to_zl_color(&get_or(style, "strokeColor", inum(-12303292)), -12303292)),
            "pressedBorderColor": inum(fcl_argb_to_zl_color(&get_or(style, "strokeColorPressed", inum(-12303292)), -12303292)),
            "borderRadius": radius_obj,
            "pressedBorderRadius": pressed_radius_obj,
        });
        result.push(json!({
            "name": name,
            "uuid": sid,
            "animateSwap": false,
            "commonStyle": true,
            "lightStyle": light,
            "darkStyle": light.clone(),
        }));
    }
    (result, mapping)
}

pub fn make_zl_button_size(base_info: &Value, absolute_as_percentage: bool, aspect: f64) -> Value {
    if to_string_v(&get_or(base_info, "sizeType", json!(""))) == "ABSOLUTE" && absolute_as_percentage {
        let screen_height_dp = 411.0f64;
        let screen_width_dp = screen_height_dp * 0.1f64.max(clamp_float(&gof_num(aspect), 16.0 / 9.0));
        let width_dp = clamp_zl_dp(&get_or(base_info, "absoluteWidth", inum(50)), 50.0);
        let height_dp = clamp_zl_dp(&get_or(base_info, "absoluteHeight", inum(50)), 50.0);
        let width_percentage =
            100.max(10000.min(py_round(width_dp / screen_width_dp * 10000.0)));
        let height_percentage =
            100.max(10000.min(py_round(height_dp / screen_height_dp * 10000.0)));
        return json!({
            "type": "percentage",
            "widthDp": py_num(width_dp),
            "heightDp": py_num(height_dp),
            "widthPercentage": inum(width_percentage),
            "heightPercentage": inum(height_percentage),
            "widthReference": "screen_width",
            "heightReference": "screen_height",
        });
    }
    let mut size_type = "percentage";
    if to_string_v(&get_or(base_info, "sizeType", json!(""))) == "ABSOLUTE" {
        size_type = "dp";
    }
    let pw = get_or_obj_ref(base_info, "percentageWidth");
    let ph = get_or_obj_ref(base_info, "percentageHeight");
    let empty = obj();
    let pw = pw.unwrap_or(&empty);
    let ph = ph.unwrap_or(&empty);
    json!({
        "type": size_type,
        "widthDp": py_num(clamp_zl_dp(&get_or(base_info, "absoluteWidth", inum(50)), 50.0)),
        "heightDp": py_num(clamp_zl_dp(&get_or(base_info, "absoluteHeight", inum(50)), 50.0)),
        "widthPercentage": inum(fcl_size_to_zl(&get_or(pw, "size", inum(50)))),
        "heightPercentage": inum(fcl_size_to_zl(&get_or(ph, "size", inum(50)))),
        "widthReference": fcl_ref_to_zl(&to_string_v(&get_or(pw, "reference", json!("")))),
        "heightReference": fcl_ref_to_zl(&to_string_v(&get_or(ph, "reference", json!("")))),
    })
}
